Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool.swift

```diff
@@ -282,49 +323,29 @@ final class HTTPConnectionPool {
     private func scheduleRequestTimeout(_ request: Request, on eventLoop: EventLoop) {
         let requestID = request.id
         let scheduled = eventLoop.scheduleTask(deadline: request.connectionDeadline) {
-            // The timer has fired. Now we need to do a couple of things:
-            //
-            // 1. Remove ourselves from the timer dictionary to not leak any data. If our
-            //    waiter entry still exists, we need to tell the state machine, that we want
-            //    to fail the request.
-            let timeoutFired = self.timerLock.withLock {
-                self._requestTimer.removeValue(forKey: requestID) != nil
-            }
-
-            // 2. If the entry did not exists anymore, we can assume that the request was
-            //    scheduled on another connection. The timer still fired anyhow because of a
-            //    race. In such a situation we don't need to do anything.
-            guard timeoutFired else { return }
-
-            // 3. Tell the state machine about the timeout
-            self.modifyStateAndRunActions {
-                $0.timeoutRequest(requestID)
+            // there might be a race between a the timeout timer and the pool scheduling the
+            // request on another thread.
+            self.modifyStateAndRunActions { stateMachine in
+                if self._requestTimer.removeValue(forKey: requestID) != nil {
+                    // The timer still exists. State Machines assumes it is alive. Inform state
+                    // machine.
+                    return stateMachine.timeoutRequest(requestID)
+                }
+                return .none
             }
         }
 
-        self.timerLock.withLockVoid {
-            assert(self._requestTimer[requestID] == nil)
-            self._requestTimer[requestID] = scheduled
-        }
+        assert(self._requestTimer[requestID] == nil)
+        self._requestTimer[requestID] = scheduled
 
         request.req.requestWasQueued(self)
     }
 
     private func cancelRequestTimeout(_ id: Request.ID) {
-        let scheduled = self.timerLock.withLock {
-            self._requestTimer.removeValue(forKey: id)
+        guard let cancelTimer = self._requestTimer.removeValue(forKey: id) else {
+            preconditionFailure("Expected to have a timer for request \(id) at this point.")
         }
-
-        scheduled?.cancel()
-    }
-
-    private func cancelRequestTimeouts(_ requests: [Request]) {
-        let scheduled = self.timerLock.withLock {
-            requests.compactMap {
-                self._requestTimer.removeValue(forKey: $0.id)
-            }
-        }
-        scheduled.forEach { $0.cancel() }
+        cancelTimer.cancel()
```

Comment:
Should we do these `cancel`s outside the lock?

Reply:
Good idea!

Reply:
Will do at a later point, tracked here: https://github.com/swift-server/async-http-client/issues/439